Add the generic operation.

ananos opened the pull request:
Accompanies vAccelRT branch feat_genop@5112a04619800

VACCEL_GEN_OP is a generic operation where the frontend has a
prototype of the form:

    function(output, input, output_len, input_len);

in order to serve as a generic abstraction for any kind of
execution.

The form of input and output parameters is completely agnostic
to the framework. vAccel just forwards a call to a specific
function implemented as a plugin. The actual arguments are unpacked
in the plugin and packed in a small wrapper template in the
frontend (not included in this patchset). For clarity, the wrapper
template should be available as an example to help individuals port
their functions to vAccel.

Signed-off-by: Anastassios Nanos <ananos@nubificus.co.uk>

ananos committed:
Add the generic operation.

Accompanies vAccelRT branch feat_genop@5112a04619800

VACCEL_GEN_OP is a generic operation where the frontend has a
prototype of the form:

    function(output, input, output_len, input_len);

in order to serve as a generic abstraction for any kind of
execution.

The form of input and output parameters is completely agnostic
to the framework. vAccel just forwards a call to a specific
function implemented as a plugin. The actual arguments are unpacked
in the plugin and packed in a small wrapper template in the
frontend (not included in this patchset). For clarity, the wrapper
template should be available as an example to help individuals port
their functions to vAccel.

Signed-off-by: Anastassios Nanos <ananos@nubificus.co.uk>

ananos committed:
testing: Enable aarch64 image building

Signed-off-by: Anastassios Nanos <ananos@nubificus.co.uk>

papazof commented on src/devices/src/virtio/crypto/request.rs:
```diff
@@ -225,6 +225,34 @@ impl Request {
+        sess: &mut vaccel_session,
+        mem: &GuestMemoryMmap
+    ) -> Result<u32> {
+        let (out_addr, out_len) = self.in_args[0];
+        let mut out_args = vec![0u8; out_len as usize];
+        mem.read(&mut out_args, out_addr).map_err(Error::GuestMemory)?;
+
+        let (in_addr, in_len) = self.out_args[1];
+        let mut in_args = vec![0u8; in_len as usize];
+        mem.read(&mut in_args, in_addr).map_err(Error::GuestMemory)?;
```
Shouldn't out_args and in_args be the other way round?

ananos committed:
Properly parse genop arguments.

Iterate over the supplied arguments and generate u32, *u8
tuples to be fed to the relevant vaccel-bindings operation.

mem.read is now zero-copy (we only pass the pointer to the
memory space residing in the guest's kernel address space).

compatible with vaccel-bindings branch feat_genop
commit: 165cc0f

Signed-off-by: Anastassios Nanos <ananos@nubificus.co.uk>
Signed-off-by: Babis Chalios <bchalios@nubificus.co.uk>

bchalios committed:
Use only vaccel_genop calls to vAccelRT

vAccelRT provides now a vaccel_genop call which operates as a dispatcher
to an arbitrary vAccel API call. It receives a set of read-only and
read-write arguments to pass to the call and the op code of the API
function to be called (as the first read-only argument).

We use this call to replace all individual calls (such as
vaccel_image_classify), to avoid frequent changes to the Firecracker
implementation. Adding a new call to the vAccel API will not require
modifications to Firecracker from now on (hopefully).

Signed-off-by: Babis Chalios <mail@bchalios.io>

bchalios committed:
ci/cd: do not use cache when building docker image

Avoid re-using cache to make sure we use the correct vaccelrt version.

Signed-off-by: Babis Chalios <mail@bchalios.io>

bchalios committed:
ci/cd: delete docker manifest before re-creating

If we don't delete the old manifest for `nubificus/fcuvm:vaccel` we
might end-up having multiple entries for the same architecture.

Signed-off-by: Babis Chalios <mail@bchalios.io>

bchalios committed:
Update pointer to vaccel-bindings crate

Branch api_cleanup was merged to master, so we can use that now.

Signed-off-by: Babis Chalios <mail@bchalios.io>
